# DS 3100 Activity Solution: Exploring the Tennessee Education Data

This notebook provides a solution to the data-frame activity using **R** and the provided `tenn2018.csv` file.

> **Note:** Place `tenn2018.csv` in a data folder that is in the same directory as this notebook before running it.

## 1. Read the data and print the first few observations

We use `read.csv()` to read the CSV file into an R **data frame** called `ten`. The `head()` function displays the first six observations by default.

In [2]:
ten <- read.csv("data/tenn2018.csv")
head(ten)

,district_number,district_name,school_number,school_name,year,subgroup,overall_subject,percent_below,percent_approaching,percent_on_track,percent_mastered,percent_on_mastered,baseline_year,percent_below_previous,percent_approaching_previous,percent_on_track_previous,percent_mastered_previous,percent_on_mastered_previous
,<int>,<chr>,<int>,<chr>,<int>,<chr>,<chr>,<dbl>,<dbl>,<dbl>,<dbl>,<dbl>,<int>,<dbl>,<dbl>,<dbl>,<dbl>,<dbl>
1,0,State of Tennessee,0,,2018,All Students,ELA,19.5,47.7,28.1,4.7,32.8,2017,20.0,45.9,28.0,6.1,34.1
2,0,State of Tennessee,0,,2018,Black/Hispanic/Native,ELA,30.1,50.5,17.5,1.9,19.4,2017,30.9,49.2,17.6,2.3,19.9
3,0,State of Tennessee,0,,2018,Economically Disadvantaged,ELA,30.4,51.2,16.7,1.7,18.4,2017,31.5,49.6,16.9,2.0,18.9
4,0,State of Tennessee,0,,2018,English Learners,ELA,30.1,51.7,16.7,1.6,18.3,2017,34.3,49.1,15.1,1.5,16.6
5,0,State of Tennessee,0,,2018,Foster,ELA,33.9,50.0,14.4,1.8,16.2,NA,NA,NA,NA,NA,NA
6,0,State of Tennessee,0,,2018,Homeless,ELA,35.8,47.9,14.8,1.5,16.3,2017,35.5,47.2,15.2,2.1,17.4


The first few rows show the structure of the dataset: each row is an observation, while each column is a variable.

## 2. Determine the number of rows and columns

`nrow()` returns the number of observations (rows), and `ncol()` returns the number of variables (columns).

In [3]:
nrow(ten)
ncol(ten)

[1] 121062

[1] 18

**Answer:** The dataset contains **121,062 rows** and **18 columns**.

## 3. Print the column names

`names()` returns the names of the variables in the data frame. The length of this vector is the number of columns.

In [8]:
cat("Column names: ")
cat(names(ten), sep = ", " , "\n")
cat("Number of columns:", length(names(ten)), "\n")

Column names: district_number, district_name, school_number, school_name, year, subgroup, overall_subject, percent_below, percent_approaching, percent_on_track, percent_mastered, percent_on_mastered, baseline_year, percent_below_previous, percent_approaching_previous, percent_on_track_previous, percent_mastered_previous, percent_on_mastered_previous, 
Number of columns: 18 


**Answer:** There are **18 variables** in the dataset. Looking into the data, there appears to be 4 columns holding categorical values.

## 4. Access the `overall_subject` column

The `$` operator is a convenient way to access a named column of a data frame. The expression below returns the complete `overall_subject` column.

In [ ]:
ten$overall_subject

[1] "ELA"            "ELA"            "ELA"            "ELA"           
    [5] "ELA"            "ELA"            "ELA"            "ELA"           
    [9] "ELA"            "ELA"            "ELA"            "ELA"           
   [13] "ELA"            "ELA"            "ELA"            "ELA"           
   [17] "Math"           "Math"           "Math"           "Math"          
   [21] "Math"           "Math"           "Math"           "Math"          
   [25] "Math"           "Math"           "Math"           "Math"          
   [29] "Math"           "Math"           "Math"           "Math"          
   [33] "Science"        "Science"        "Science"        "Science"       
   [37] "Science"        "Science"        "Science"        "Science"       
   [41] "Science"        "Science"        "Science"        "Science"       
   [45] "Science"        "Science"        "Science"        "Science"       
   [49] "Social Studies" "Social Studies" "Social Studies" "Social Studies"
   [53] "Social Studies" "Social Studies" "Social Studies" "Social Studies"
   [57] "Social Studies" "Social Studies" "Social Studies" "Social Studies"
   [61] "Social Studies" "Social Studies" "Social Studies" "Social Studies"
   [65] "ELA"            "ELA"            "ELA"            "ELA"           
   [69] "ELA"            "ELA"            "ELA"            "ELA"           
   [73] "ELA"            "ELA"            "ELA"            "ELA"           
   [77] "ELA"            "ELA"            "ELA"            "ELA"           
   [81] "Math"           "Math"           "Math"           "Math"          
   [85] "Math"           "Math"           "Math"           "Math"          
   [89] "Math"           "Math"           "Math"           "Math"          
   [93] "Math"           "Math"           "Math"           "Math"          
   [97] "Science"        "Science"        "Science"        "Science"       
  [101] "Science"        "Science"        "Science"        "Science"       
  [105] "Science"        "Science"        "Science"        "Science"       
  [109] "Science"        "Science"        "Science"        "Science"       
  [113] "Social Studies" "Social Studies" "Social Studies" "Social Studies"
  [117] "Social Studies" "Social Studies" "Social Studies" "Social Studies"
  [121] "Social Studies" "Social Studies" "Social Studies" "Social Studies"
  [125] "Social Studies" "Social Studies" "Social Studies" "Social Studies"
  [129] "ELA"            "ELA"            "ELA"            "ELA"           
  [133] "ELA"            "ELA"            "ELA"            "ELA"           
  [137] "ELA"            "ELA"            "ELA"            "ELA"           
  [141] "ELA"            "ELA"            "ELA"            "ELA"           
  [145] "Math"           "Math"           "Math"           "Math"          
  [149] "Math"           "Math"           "Math"           "Math"          
  [153] "Math"           "Math"           "Math"           "Math"          
  [157] "Math"           "Math"           "Math"           "Math"          
  [161] "Science"        "Science"        "Science"        "Science"       
  [165] "Science"        "Science"        "Science"        "Science"       
  [169] "Science"        "Science"        "Science"        "Science"       
  [173] "Science"        "Science"        "Science"        "Science"       
  [177] "Social Studies" "Social Studies" "Social Studies" "Social Studies"
  [181] "Social Studies" "Social Studies" "Social Studies" "Social Studies"
  [185] "Social Studies" "Social Studies" "Social Studies" "Social Studies"
  [189] "Social Studies" "Social Studies" "Social Studies" "Social Studies"
  [193] "ELA"            "ELA"            "ELA"            "ELA"           
  [197] "ELA"            "ELA"            "ELA"            "ELA"           
  [201] "ELA"            "ELA"            "ELA"            "ELA"           
  [205] "ELA"            "ELA"            "ELA"            "ELA"           
  [209] "Math"           "Math"           "Math"    

## 5. Access `district_name` for the first observation

A data frame can be indexed using `[row, column]`. R uses **1-based indexing**, so the first observation is row `1`.

In [ ]:
ten[1, "district_name"]

[1] "State of Tennessee"

**Answer:** `"State of Tennessee"`

## 6. Access `percent_below` for the first observation

In [ ]:
ten[1, "percent_below"]

[1] 19.5

**Answer:** `19.5`

## 7. How many columns contain missing values?

First, `is.na(ten)` creates a logical table indicating where values are missing.

`colSums(is.na(ten))` counts the missing values in each column. We then check which columns have **at least one** missing value and count them.

In [9]:
missing_by_column <- colSums(is.na(ten))

cat("Number of columns with missing values:", sum(missing_by_column > 0), "\n")

Number of columns with missing values: 11 


**Answer:** **15 columns** contain at least one missing value.

The other **3 columns**—`district_number`, `school_number`, and `year`—have no missing values.